crates/ruff_linter/src/rules/flake8_pyi/rules/unnecessary_literal_union.rs

```diff
@@ -32,43 +34,163 @@ pub struct UnnecessaryLiteralUnion {
+    }
+}
+
+fn concatenate_bin_ors(exprs: Vec<&Expr>) -> Expr {
+    let mut exprs = exprs.into_iter();
+    let first = exprs.next().unwrap();
+    exprs.fold((*first).clone(), |acc, expr| {
+        Expr::BinOp(ast::ExprBinOp {
+            left: Box::new(acc),
+            op: Operator::BitOr,
+            right: Box::new((*expr).clone()),
+            range: TextRange::default(),
+        })
+    })
+}
+
+fn make_union(subscript: &ExprSubscript, exprs: Vec<&Expr>) -> Expr {
+    Expr::Subscript(ast::ExprSubscript {
+        value: subscript.value.clone(),
+        slice: Box::new(Expr::Tuple(ast::ExprTuple {
+            elts: exprs.into_iter().map(|expr| (*expr).clone()).collect(),
+            range: TextRange::default(),
+            ctx: ast::ExprContext::Load,
+        })),
+        range: TextRange::default(),
+        ctx: ast::ExprContext::Load,
+    })
+}
+
+fn make_literal_expr(subscript: Option<Expr>, exprs: Vec<&Expr>) -> Expr {
+    let use_subscript = if let subscript @ Some(_) = subscript {
+        subscript.unwrap().clone()
+    } else {
+        Expr::Name(ast::ExprName {
+            id: "Literal".to_string(),
+            range: TextRange::default(),
+            ctx: ast::ExprContext::Load,
+        })
+    };
+    Expr::Subscript(ast::ExprSubscript {
+        value: Box::new(use_subscript),
+        slice: Box::new(Expr::Tuple(ast::ExprTuple {
+            elts: exprs.into_iter().map(|expr| (*expr).clone()).collect(),
+            range: TextRange::default(),
+            ctx: ast::ExprContext::Load,
+        })),
+        range: TextRange::default(),
+        ctx: ast::ExprContext::Load,
+    })
 }
 
 /// PYI030
 pub(crate) fn unnecessary_literal_union<'a>(checker: &mut Checker, expr: &'a Expr) {
     let mut literal_exprs = Vec::new();
+    let mut other_exprs = Vec::new();
 
-    // Adds a member to `literal_exprs` if it is a `Literal` annotation
+    // for the sake of consistency and correctness, we'll use the first Literal subscript attribute
+    let mut literal_subscript = None;
+    let mut total_literals = 0;
+
+    // Split members into `literal_exprs` if they are a `Literal` annotation  and `other_exprs` otherwise
     let mut collect_literal_expr = |expr: &'a Expr, _| {
         if let Expr::Subscript(ast::ExprSubscript { value, slice, .. }) = expr {
             if checker.semantic().match_typing_expr(value, "Literal") {
-                literal_exprs.push(slice);
+                total_literals += 1;
+
+                // flatten already-unioned literals to later union again
+                if let Expr::Tuple(ast::ExprTuple {
+                    elts,
+                    range: _,
+                    ctx: _,
+                }) = slice.as_ref()
+                {
+                    for expr in elts {
+                        if literal_subscript.is_none() {
+                            literal_subscript = Some(*value.clone());
+                        }
```

Comment:
It looks like this could be moved outside of the check for a tuple and deduplicated

Reply:
woop, silly mistake, will fix.